### PR TITLE
Upgrade to Python 3.8.3

### PR DESCRIPTION
**Pull request by dand-oss**
NOW Python 3.8.6

Python 3.8.3 - using this on Windows and Linux in release and debug

Lotta work!

they renamed libpython3.8m.so to leave off the ABI, so now libpython3.8.so

On Windows they broke LoadLibrary() for C extensions and ctypes - it ignores path!
So if a c-extsionsion use dlls, you need to copy them into the directory with the pyd.

https://bugs.python.org/issue36085

1 years later and conda still hasn't released the 3.8 version of many packages with C code...

https://github.com/conda-forge/python-feedstock/issues/307

The python build on windows now downloads a bunch of binary libraries via PCBuild/get_externals.py to build its C extensions. 

However, you can build zlib / lexpat / ffi /sqlite using cmake on your own Windows system.  We could add download of "externals" in cmake if someone wants to, or mabye use CMake externals.

Requires FFI https://github.com/winlibs/libffi/pull/4 on Windows
Requires ZLIB http://github.com/madler/zlib.git  on Windows
Requires Expat https://github.com/dand-oss/libexpat/tree/top-level-cmake on Windows
Requires SSL https://github.com/dand-oss/openssl-cmake/commits/cmake-cleanup on Windows
Helpful sqlite https://github.com:alex85k/sqlite3-cmake.git on Windows

Not a lot of backporting, just wanted to "make it work"

### CODE REVIEW

**Comment by jcfr** [comment]
@dand-oss  Thanks for your help with  this :pray:  Supporting newer version of CPython  would be great.

I currently have very limited bandwidth  to help reviewing this change.

To help, would you (or anyone else) have financial  resources to allocate to support this effort ?

File: cmake/libpython/CMakeLists.txt [hunk 1]
```diff
@@ -24,6 +27,8 @@ if(UNIX)
             PYTHONPATH="${PYTHONPATH}"
       )
 elseif(WIN32)
+    # needed for python 3.5
+    add_definitions(-DPY3_DLLNAME="python3${CMAKE_DEBUG_POSTFIX}")
```
Comment (gongminmin) [inline cmake/libpython/CMakeLists.txt:31]: This line makes PY3_DLL_NAME has debug postfix even in release configurations.

**Comment by anki-code** [comment]
Hello @dand-oss! Many thanks to you work! 

I used your repo to get `libpython3.8.a` file (and I got this! thanks!) but the make command wasn't successfully ended and I just want to understand why. Could you please comment the error? Here is the example with output:

<details>

```bash
docker run -it --rm alpine
apk add git cmake make gcc musl-dev
mkdir -p ~/scratch && cd ~/scratch
git clone --depth 1 -b upgrade-py383 https://github.com/dand-oss/python-cmake-buildsystem
mkdir -p python-build && mkdir -p python-install
cd python-build
cmake -DBUILD_EXTENSIONS_AS_BUILTIN=ON -DBUILTIN_OSSAUDIODEV=OFF -DENABLE_OSSAUDIODEV=OFF -DENABLE_LINUXAUDIODEV=OFF -DBUILTIN_LINUXAUDIODEV=OFF -DENABLE_AUDIOOP=OFF -DBUILTIN_AUDIOOP=OFF -DCMAKE_INSTALL_PREFIX:PATH=${HOME}/scratch/python-install ../python-cmake-buildsystem
make -j10
```
The last lines:
```
...
...
[ 97%] Building C object CMakeBuild/libpython/CMakeFiles/libpython-static.dir/root/scratch/Python-3.8.6/Modules/sha1module.c.o
[ 97%] Building C object CMakeBuild/libpython/CMakeFiles/libpython-static.dir/root/scratch/Python-3.8.6/Modules/sha256module.c.o
[ 97%] Building C object CMakeBuild/libpython/CMakeFiles/libpython-static.dir/root/scratch/Python-3.8.6/Modules/sha512module.c.o
[ 97%] Building C object CMakeBuild/libpython/CMakeFiles/libpython-static.dir/root/scratch/Python-3.8.6/Modules/binascii.c.o
[ 98%] Building C object CMakeBuild/libpython/CMakeFiles/libpython-static.dir/root/scratch/Python-3.8.6/Python/frozen.c.o
[ 98%] Linking C static library ../../lib/libpython3.8.a
[ 98%] Built target libpython-static
Scanning dependencies of target python
[ 98%] Building C object CMakeBuild/python/CMakeFiles/python.dir/root/scratch/Python-3.8.6/Programs/python.c.o
In file included from /root/scratch/Python-3.8.6/Programs/python.c:4:
/root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pylifecycle.h:8:4: error: #error "this header requires Py_BUILD_CORE define"
    8 | #  error "this header requires Py_BUILD_CORE define"
      |    ^~~~~
In file included from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pylifecycle.h:11,
                 from /root/scratch/Python-3.8.6/Programs/python.c:4:
/root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_initconfig.h:8:4: error: #error "this header requires Py_BUILD_CORE define"
    8 | #  error "this header requires Py_BUILD_CORE define"
      |    ^~~~~
In file included from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_initconfig.h:11,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pylifecycle.h:11,
                 from /root/scratch/Python-3.8.6/Programs/python.c:4:
/root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pystate.h:8:4: error: #error "this header requires Py_BUILD_CORE define"
    8 | #  error "this header requires Py_BUILD_CORE define"
      |    ^~~~~
In file included from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pystate.h:17,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_initconfig.h:11,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pylifecycle.h:11,
                 from /root/scratch/Python-3.8.6/Programs/python.c:4:
/root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_gil.h:8:4: error: #error "this header requires Py_BUILD_CORE define"
    8 | #  error "this header requires Py_BUILD_CORE define"
      |    ^~~~~
In file included from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_gil.h:11,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pystate.h:17,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_initconfig.h:11,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pylifecycle.h:11,
                 from /root/scratch/Python-3.8.6/Programs/python.c:4:
/root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_condvar.h:5:4: error: #error "this header requires Py_BUILD_CORE define"
    5 | #  error "this header requires Py_BUILD_CORE define"
      |    ^~~~~
In file included from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_gil.h:12,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pystate.h:17,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_initconfig.h:11,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pylifecycle.h:11,
                 from /root/scratch/Python-3.8.6/Programs/python.c:4:
/root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_atomic.h:8:4: error: #error "this header requires Py_BUILD_CORE define"
    8 | #  error "this header requires Py_BUILD_CORE define"
      |    ^~~~~
In file included from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pystate.h:18,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_initconfig.h:11,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pylifecycle.h:11,
                 from /root/scratch/Python-3.8.6/Programs/python.c:4:
/root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pathconfig.h:8:4: error: #error "this header requires Py_BUILD_CORE define"
    8 | #  error "this header requires Py_BUILD_CORE define"
      |    ^~~~~
In file included from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pystate.h:19,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_initconfig.h:11,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pylifecycle.h:11,
                 from /root/scratch/Python-3.8.6/Programs/python.c:4:
/root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pymem.h:8:4: error: #error "this header requires Py_BUILD_CORE define"
    8 | #  error "this header requires Py_BUILD_CORE define"
      |    ^~~~~
In file included from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pystate.h:20,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_initconfig.h:11,
                 from /root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_pylifecycle.h:11,
                 from /root/scratch/Python-3.8.6/Programs/python.c:4:
/root/scratch/python-build/../Python-3.8.6/Include/internal/pycore_warnings.h:8:4: error: #error "this header requires Py_BUILD_CORE define"
    8 | #  error "this header requires Py_BUILD_CORE define"
      |    ^~~~~
make[2]: *** [CMakeBuild/python/CMakeFiles/python.dir/build.make:63: CMakeBuild/python/CMakeFiles/python.dir/root/scratch/Python-3.8.6/Programs/python.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1257: CMakeBuild/python/CMakeFiles/python.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

</details>

Thanks!

**Comment by dand-oss** [comment]
> I used your PR-branch to get `libpython3.8.a` file (and I got this! thanks!) but the make command wasn't successfully ended and I just want to understand why. Could you please comment the error? Here is the example with output:

OMG - Alpine!

suggest stick with "glibc" over "musl"
ex, if we get through python then every c extension will have to work with musl

If you are going for smallest size, can you use debian slim instead?

what is what I use, will be easier to help you - say in setting up a multi-stage build docker

IAC  I will dive in more later

**Comment by anki-code** [comment]
@dand-oss Yeah, Alpine because My final goal is to create statically linked binary with musl like I did in [xonsh-portable](https://github.com/anki-code/xonsh-portable-binary/blob/main/xonsh-portable-musl-alpine.Dockerfile). 
I'm not sure that the same could be achieved with glibc. Or it could be?

**Comment by anki-code** [comment]
When I run the binary file that compiled with `libpython3.8.a` I've got the error:

...
  File "/tmp/asyncio/unix_events.py", line 174, in _check_signal
AttributeError: module 'signal' has no attribute 'valid_signals'
```

Stack:

<details>

```
Traceback (most recent call last):
  File "/tmp/prompt_toolkit/application/application.py", line 1317, in attach_winch_signal_handler
  File "/tmp/asyncio/unix_events.py", line 88, in add_signal_handler
  File "/tmp/asyncio/unix_events.py", line 174, in _check_signal
AttributeError: module 'signal' has no attribute 'valid_signals'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/xonsh/main.py", line 425, in main
  File "/tmp/xonsh/main.py", line 463, in main_xonsh
  File "/tmp/xonsh/ptk_shell/shell.py", line 338, in cmdloop
  File "/tmp/xonsh/ptk_shell/shell.py", line 307, in singleline
  File "/tmp/prompt_toolkit/shortcuts/prompt.py", line 1020, in prompt
  File "/tmp/prompt_toolkit/application/application.py", line 887, in run
  File "/tmp/asyncio/base_events.py", line 616, in run_until_complete
  File "/tmp/prompt_toolkit/application/application.py", line 816, in run_async
  File "/tmp/prompt_toolkit/application/application.py", line 792, in _run_async2
  File "/tmp/prompt_toolkit/application/application.py", line 721, in _run_async
  File "/tmp/contextlib.py", line 113, in __enter__
  File "/tmp/prompt_toolkit/application/application.py", line 1321, in attach_winch_signal_handler
  File "/tmp/asyncio/unix_events.py", line 138, in remove_signal_handler
  File "/tmp/asyncio/unix_events.py", line 174, in _check_signal
AttributeError: module 'signal' has no attribute 'valid_signals'
```

</details>

The [docs of signal](https://docs.python.org/3.8/library/signal.html) says: valid_signails is new in version 3.8.

Does it related to issues in the python build system?

**Comment by jamesobutler** [comment]
Hi @dand-oss!

Does this branch represent your latest work, or is the latest within https://github.com/dand-oss/python-cmake-buildsystem/tree/3.9.0 or https://github.com/dand-oss/python-cmake-buildsystem/tree/master? Have you successfully added python 3.9 support?

**Comment by fbordignon** [comment]
@dand-oss thanks for the contributions. Can you update us on the state of your branches? This update is very much anticipated. 
I've tried to build https://github.com/dand-oss/python-cmake-buildsystem/tree/3.9.0
But I get an error that I do not have the include asv_cmake_defaults.
I am not sure what it is and how to work around it. Can you elucidate me?

Thank you!

**Comment by jcfr** [comment]
@dand-oss Working on integrating contributions related to support for python 3.7, 3.8 and 3.9, I noticed that the email you used to commit your changes is not associated with your account anymore. To have your contribution accounted in contributor reports, which email you would like us to use ? 

If you prefer to keep your email private, you could also follow instructions here: https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-github-user-account/managing-email-preferences/setting-your-commit-email-address and let us know which "no-reply email" should be used when re-basing your changes when referencing your using `Co-authored-by: ....`

Thanks :pray:

File: cmake/libpython/CMakeLists.txt (hunk 1 above)
Comment (jcfr) [reply cmake/libpython/CMakeLists.txt:31]: @gongminmin Thanks for pointing this out :pray:  For future reference, the corresponding change was integrated as part of https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/pull/286

**Comment by jcfr** [comment]
Closing. This pull-request is superseded by https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/pull/289 and https://github.com/python-cmake-buildsystem/python-cmake-buildsystem/pull/292

**Pull request closed without merging**